File: apps/web/src/types/power-entity.ts
/**
 * PowerAtlas data contract.
 *
 * This is the exact JSON shape the Phase 2 FastAPI backend will serve; the
 * Phase 1 mock files in `src/data/mock` already conform to it, so swapping
 * `services/mockDataLoader.ts` for a real `apiClient.ts` is a one-file
 * change. `status` / `confidence` / `sources` exist now because the future
 * review workflow (draft -> single-admin approval -> published) needs them.
 */

export type ConfidenceLevel = 'high' | 'medium' | 'low'

export type ReviewStatus = 'draft' | 'published'

export type PowerDimension = 'official' | 'hidden'

/**
 * The other axis of the national card: a collective actor (a party, an
 * institution, a documented organization) versus a single person (an office
 * holder, a businessperson). Cross with `dimension` to get the four sections.
 */
export type EntityClass = 'group' | 'individual'

export type EntityKind =
  | 'office'
  | 'institution'
  | 'organization'
  | 'faction'
  | 'movement'
  | 'economic-bloc'
  | 'person'

/**
 * Decomposition of `score` into the three pillars of the power scale
 * (0-100 each). See lib/powerScore.ts and docs/power-scale.md. When present,
 * `score` should equal `powerScore(power)`.
 */
export interface PowerBreakdown {
  capital: number
  authority: number
  influence: number
}

export interface SourceCitation {
  id: string
  label: string
  url?: string
  publishedAt?: string
  note?: string
}

export interface PowerEntity {
  id: string
  name: string
  kind: EntityKind
  dimension: PowerDimension
  /** 0-100 influence index. Phase 1: dummy placeholder values. */
  score: number
  /** Score change since the previous assessment cycle. */
  delta: number
  confidence: ConfidenceLevel
  status: ReviewStatus
  sources: SourceCitation[]
  note?: string
  /** Group vs individual — the second axis of the national 4-section card. */
  entityClass?: EntityClass
  /** The three-pillar decomposition behind `score`. */
  power?: PowerBreakdown
  /** Party sigla for an individual office holder (badge). */
  party?: string
  /** Human-readable role/title ("Senador", "Governador", "Empresário"). */
  role?: string
}

export interface RegionCapital {
  name: string
  /** [longitude, latitude] — anchors the deck.gl column/arc layers. */
  coordinates: [number, number]
}

export interface PowerRegion {
  /** 'BR' for the country, otherwise the UF sigla (joins GeoJSON `UF`). */
  id: string
  name: string
  kind: 'country' | 'state'
  capital: RegionCapital
  updatedAt: string
  official: PowerEntity[]
  hidden: PowerEntity[]
}

export interface InfluenceLink {
  id: string
  /** Region ids ('BR' or UF sigla). */
  from: string
  to: string
  /** 0-1 relative strength; drives arc width/opacity. */
  strength: number
  dimension: PowerDimension
  label?: string
}

export interface AmbientSignal {
  /** [longitude, latitude] */
  coordinates: [number, number]
  /** 0-1 heatmap weight. */
  weight: number
}

export interface RegionPowerData {
  schemaVersion: 1
  generatedAt: string
  /** Rendered permanently in the UI while data is simulated. */
  disclaimer: string
  regions: PowerRegion[]
  links: InfluenceLink[]
  ambientSignals: AmbientSignal[]
}

File: apps/api/src/models/power_entity.py
"""Pydantic mirror of the TypeScript data contract.

Field-for-field port of ``apps/web/src/types/power-entity.ts``. Field names are
snake_case in Python and serialized as camelCase via ``to_camel`` aliases, so
the JSON payload is byte-compatible with what the web ``mockDataLoader`` builds
today. ``extra="forbid"`` makes any drift between the mock JSON and this schema
fail loudly instead of passing silently.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

ConfidenceLevel = Literal["high", "medium", "low"]
ReviewStatus = Literal["draft", "published"]
PowerDimension = Literal["official", "hidden"]
EntityKind = Literal[
    "office",
    "institution",
    "organization",
    "faction",
    "movement",
    "economic-bloc",
]


class _Base(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="forbid",
    )


class SourceCitation(_Base):
    id: str
    label: str
    url: str | None = None
    published_at: str | None = None
    note: str | None = None


class PowerEntity(_Base):
    id: str
    name: str
    kind: EntityKind
    dimension: PowerDimension
    score: int
    delta: int
    confidence: ConfidenceLevel
    status: ReviewStatus
    sources: list[SourceCitation]
    note: str | None = None


class RegionCapital(_Base):
    name: str
    coordinates: tuple[float, float]


class PowerRegion(_Base):
    id: str
    name: str
    kind: Literal["country", "state"]
    capital: RegionCapital
    updated_at: str
    official: list[PowerEntity]
    hidden: list[PowerEntity]


class InfluenceLink(_Base):
    id: str
    from_: str = Field(alias="from")
    to: str
    strength: float
    dimension: PowerDimension
    label: str | None = None


class AmbientSignal(_Base):
    coordinates: tuple[float, float]
    weight: float


class RegionPowerData(_Base):
    schema_version: Literal[1]
    generated_at: str
    disclaimer: str
    regions: list[PowerRegion]
    links: list[InfluenceLink]
    ambient_signals: list[AmbientSignal]
